design/anoncreds.md

```diff
@@ -0,0 +1,507 @@
+anoncerds protocol:
+
+* Each Credential must have a reserved mandatory attribute: `issuanceTime`.
+    * It's set by the Issuer to specify the time of Issuance.
+    * It's needed to fulfill Requirements 5.
+* This attribute can be considered as `m3` special attribute (`m1` is master secret, `m2` is credential context, `m3` is issuance time).
+* Since the Issuer may be malicious (if keys were compromised already), then 
+a proof that `issuanceTime` is really the current time and not the time from the past is needed.
+    * We can use our blockchain to prepare such a proof.
+    * Issuer signs (by his Cred Def's public key) the `issuanceTime` and sends it to the pool.
+    * Each node verifies that `issuanceTime` is not less that the current one, and signs the result with BLS key.
+    * Each node then sends the signed result to the Issuer (no need to write anything to the ledger).
+    * The issuer prepares a BLS multi-signature (making sure that there is a consensus)
+    and adds the BLS-signed proof of the `issuanceTime` to the credential.
+    * The verifier will then use the proof to make sure that the `issuanceTime` is really the correct one.
+* The `issuanceTime` needs to be verified in each proof.
+    * The Verifier should use Predicates (instead of disclosing) for the value of `issuanceTime`
+    to avoid correlation. 
+    * It's possible also to disclose `issuanceTime`, but we don't force it.
+    * If it's not disclosed and not verified as a Predicate, then there is a chance the the proof verification will fail because 
+of key rotations, since the latest keys will be used.
+
+
+
+### SCHEMA
+
+#### SCHEMA txn
+```
+{
+    "data": {
+        "uuid":"GEzcdDLhCpGCYRHW82kjHd",
+        "attrNames": ["undergrad","last_name","first_name","birth_date","postgrad","expiry_date"],
+        "name":"Degree",
+        "version":"1.0",
+    },
+    
+    "reqMetadata": {
+        "issuerDid":"L5AD5g65TDQr1PPHHRoiGf",
+        .....
+    },
+    
+....
+}
+```
+`uuid` is Schema's UUID. It's different from `issuerDid`.
+
+#### Restrictions
+
+* Existing Schema (identified by the Schema `uuid`) can be modified/changed/evolved.
+* Only the `issuerDid` who created the Schema can modify it (that is we need to keep the ownership).
+* It's not possible to create multiple entities with the same `uuid` (so, it's unique within the ledger). 
+
+#### State
+
+We need to have two records for Schema in State Trie in order to have
+1. Simple referencing of Schemas in the protocol (by Schema DID)
+1. Requirements 8
+
+Record 1:
+* key: `schemaIssuerDid | SchemaMarker | schemaName | schemaVersion | schemaUUID` 
+* value: aggregated txn data
+
+Record 2:
+* key: `schemaUUID`
+* value: Record 1 key
+
+
+#### GET_SCHEMA
+```
+{
+    'data': {
+        'uuid': 'GEzcdDLhCpGCYRHW82kjHd',
+    },
+...
+}
+```
+1. Lookup State Trie to get `schemaIssuerDid | SchemaMarker |schemaName | schemaVersion | schemaUUID` by `uuid`
```

Comment:
Use a cache to lookup the key in the trie and use that key to get data with proof from the trie.

Reply:
+1